be/src/vec/exec/scan/vscan_node.cpp

```diff
@@ -868,6 +880,193 @@ Status VScanNode::_normalize_noneq_binary_predicate(VExpr* expr, VExprContext* e
+        }
+
+        if (ne_pdt == PushDownType::ACCEPTABLE) {
+            bool is_fixed_range = range.is_fixed_value_range();
+            auto not_in_range =
+                    ColumnValueRange<T>::create_empty_column_value_range(range.column_name());
+            auto fn_name = std::string("");
+            if (value.data != nullptr) {
+                auto fn_name = std::string("");
+                if constexpr (T == TYPE_CHAR || T == TYPE_VARCHAR || T == TYPE_STRING ||
+                              T == TYPE_HLL) {
+                    auto val = StringValue(value.data, value.size);
+                    if (is_fixed_range) {
+                        RETURN_IF_ERROR(_change_value_range<true>(
+                                range, reinterpret_cast<void*>(&val),
+                                ColumnValueRange<T>::remove_fixed_value_range, fn_name));
+                    } else {
+                        RETURN_IF_ERROR(_change_value_range<true>(
+                                not_in_range, reinterpret_cast<void*>(&val),
+                                ColumnValueRange<T>::add_fixed_value_range, fn_name));
+                    }
+                } else {
+                    if (is_fixed_range) {
+                        RETURN_IF_ERROR(_change_value_range<true>(
+                                range, reinterpret_cast<void*>(const_cast<char*>(value.data)),
+                                ColumnValueRange<T>::remove_fixed_value_range, fn_name));
+                    } else {
+                        RETURN_IF_ERROR(_change_value_range<true>(
+                                not_in_range,
+                                reinterpret_cast<void*>(const_cast<char*>(value.data)),
+                                ColumnValueRange<T>::add_fixed_value_range, fn_name));
+                    }
+                }
+            }
+
+            if (is_fixed_range ||
+                not_in_range.get_fixed_value_size() <= _max_pushdown_conditions_per_column) {
+                if (!is_fixed_range) {
+                    _not_in_value_ranges.push_back(not_in_range);
+                }
+                *pdt = ne_pdt;
+            }
+            range.set_compound_type(compound_type);
+        }
+
+        if (noneq_pdt == PushDownType::ACCEPTABLE) {
+            const std::string& fn_name =
+                    reinterpret_cast<VectorizedFnCall*>(expr)->fn().name.function_name;
+
+            // where A = nullptr should return empty result set
+            if (value.data != nullptr) {
+                if constexpr (T == TYPE_CHAR || T == TYPE_VARCHAR || T == TYPE_STRING ||
+                              T == TYPE_HLL) {
+                    auto val = StringValue(value.data, value.size);
+                    RETURN_IF_ERROR(_change_value_range<false>(
+                            range, reinterpret_cast<void*>(&val),
+                            ColumnValueRange<T>::add_compound_value_range, fn_name,
```

Comment:
why use special add_compound_value_range here and use normal add_fixed_value_range in other branches?

Reply:
The original intention is to reuse the fixed value code, but this may not be necessary. In the latest code, the fixed value and range value in compound conditions except leaf node of and node, have been uniformly stored in `_compound_values`